sled-agent/src/sim/collection.rs

```diff
@@ -536,91 +524,34 @@ mod test {
         assert!(!rnext.run_state.is_stopped());
         rprev = rnext;
 
+        // Propolis publishes its own transition to Stopping before it publishes
+        // Stopped.
```

Comment:
If we're going to just delegate to the propolis state machine, should we bother having sled agent publish a `Stopping` (and `Rebooting`) transition as well?

Reply:
This tore me up a little bit while I was writing this. I ended up keeping this behavior to (try to) match what the old code did, but I could be persuaded to change it.

---

As background for anyone who chances across this discussion, on a stop/reboot request issued to Nexus, what happens is that

1. Nexus asks sled agent to stop the instance
2. Sled agent asks Propolis to stop the VM (or reboot it, but for now we'll focus on stopping)
3. Propolis accepts that request and then (internally) queues a "stop" action on the VM
4. Control returns to sled agent, which *immediately* changes the instance state to "Stopping" without waiting for Propolis to start processing the request
5. Sled agent returns this updated instance state to Nexus
6. Nexus writes the state returned in (4) back to CRDB
7. Nexus refetches the instance record and returns it from the stop API

The nice thing about step 4 is that it guarantees that someone has published "state = Stopping" to the database before the instance is refetched. If we don't (try to) update the record synchronously in this path, and Propolis doesn't pick up the stop request and publish a state change quickly enough, then step 7 will return that the instance is running. That seems... odd? I called `oxide instance stop`, and the command succeeded, but the instance is still running? Hmm.

That said, this can still behave strangely: if Propolis *does* publish a state change between step 6 and step 7, then the refetched instance will have whatever state Propolis published, which might not have anything at all to do with what the user requested (e.g. if Propolis handles a queued reboot request in that window). And in that sense it's most *consistent* just to cut out step 4 and say "whatever Propolis publishes is the current state, no ifs/ands/buts" instead of having this kind of edge case.

---

All told: my preference is to keep publishing the "present tense" states from sled agent at this point, just to try to match the old sled agent behavior as much as possible. But there's an interesting API design question here that's probably worth discussing further. I can file an issue for it.